File: lib/models/kajian_session.dart
import 'kajian_note.dart';
import 'transcript_segment.dart';

/// Lifecycle status of a kajian session.
enum SessionStatus {
  recording,
  recorded,
  transcribing,
  transcribed,
  summarizing,
  completed,
  error,
}

extension SessionStatusLabel on SessionStatus {
  String get label => switch (this) {
        SessionStatus.recording => 'Recording',
        SessionStatus.recorded => 'Recorded',
        SessionStatus.transcribing => 'Transcribing…',
        SessionStatus.transcribed => 'Transcribed',
        SessionStatus.summarizing => 'Generating notes…',
        SessionStatus.completed => 'Completed',
        SessionStatus.error => 'Error',
      };

  bool get isBusy =>
      this == SessionStatus.transcribing || this == SessionStatus.summarizing;
}

/// A recorded kajian: audio + transcript + AI notes + metadata.
class KajianSession {
  final String id;
  final String title;
  final String? speaker; // ustadz / lecturer
  final String? location; // masjid / venue
  final DateTime createdAt;
  final int durationMs;

  /// Local file path to the recorded audio (may be null if discarded).
  final String? audioFilePath;

  /// BCP-47 / locale id used for transcription, e.g. "id_ID".
  final String localeId;

  final List<TranscriptSegment> transcript;
  final KajianNote? note;
  final SessionStatus status;

  const KajianSession({
    required this.id,
    required this.title,
    this.speaker,
    this.location,
    required this.createdAt,
    this.durationMs = 0,
    this.audioFilePath,
    required this.localeId,
    this.transcript = const [],
    this.note,
    this.status = SessionStatus.recorded,
  });

  /// Full transcript joined into a single plain-text string.
  String get plainTranscript =>
      transcript.map((s) => s.text.trim()).where((t) => t.isNotEmpty).join(' ');

  bool get hasTranscript => transcript.isNotEmpty;
  bool get hasNotes => note != null;

  KajianSession copyWith({
    String? title,
    String? speaker,
    String? location,
    int? durationMs,
    String? audioFilePath,
    String? localeId,
    List<TranscriptSegment>? transcript,
    KajianNote? note,
    SessionStatus? status,
  }) {
    return KajianSession(
      id: id,
      title: title ?? this.title,
      speaker: speaker ?? this.speaker,
      location: location ?? this.location,
      createdAt: createdAt,
      durationMs: durationMs ?? this.durationMs,
      audioFilePath: audioFilePath ?? this.audioFilePath,
      localeId: localeId ?? this.localeId,
      transcript: transcript ?? this.transcript,
      note: note ?? this.note,
      status: status ?? this.status,
    );
  }

  Map<String, dynamic> toJson() => {
        'id': id,
        'title': title,
        'speaker': speaker,
        'location': location,
        'createdAt': createdAt.toIso8601String(),
        'durationMs': durationMs,
        'audioFilePath': audioFilePath,
        'localeId': localeId,
        'transcript': transcript.map((s) => s.toJson()).toList(),
        'note': note?.toJson(),
        'status': status.name,
      };

  factory KajianSession.fromJson(Map<String, dynamic> json) {
    return KajianSession(
      id: json['id'] as String,
      title: json['title'] as String? ?? 'Untitled Kajian',
      speaker: json['speaker'] as String?,
      location: json['location'] as String?,
      createdAt:
          DateTime.tryParse(json['createdAt'] as String? ?? '') ??
              DateTime.now(),
      durationMs: (json['durationMs'] as num?)?.toInt() ?? 0,
      audioFilePath: json['audioFilePath'] as String?,
      localeId: json['localeId'] as String? ?? 'id_ID',
      transcript: (json['transcript'] as List?)
              ?.map((e) =>
                  TranscriptSegment.fromJson(e as Map<String, dynamic>))
              .toList() ??
          const [],
      note: json['note'] == null
          ? null
          : KajianNote.fromJson(json['note'] as Map<String, dynamic>),
      status: SessionStatus.values.firstWhere(
        (s) => s.name == json['status'],
        orElse: () => SessionStatus.recorded,
      ),
    );
  }
}

File: backend-core/app/models/kajian_session.py
"""Mirrors the Flutter app's lib/models/kajian_session.dart exactly, so the
API layer can serialize/deserialize without a translation step.
"""

from __future__ import annotations

import enum
import uuid
from datetime import datetime

from sqlalchemy import ARRAY, DateTime, Enum, Float, ForeignKey, Integer, String
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import Mapped, mapped_column, relationship

from ..db import Base


class SessionStatus(str, enum.Enum):
    recording = "recording"
    recorded = "recorded"
    transcribing = "transcribing"
    transcribed = "transcribed"
    summarizing = "summarizing"
    completed = "completed"
    error = "error"


class KajianSession(Base):
    __tablename__ = "kajian_sessions"

    # Client-generated UUID (the app already mints one on recording start),
    # kept as the primary key rather than a server-generated id, so the app
    # can reference a session before it's ever synced.
    id: Mapped[str] = mapped_column(String, primary_key=True)
    user_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("users.id", ondelete="CASCADE"), index=True
    )

    title: Mapped[str] = mapped_column(String)
    # User-typed label, entered before/at recording start — kept even after
    # speaker_id is set below, since it's the original human-entered value
    # and the confirmed profile's name could later be renamed independently.
    speaker: Mapped[str | None] = mapped_column(String, nullable=True)
    # Links to a confirmed voice profile (see models/speaker.py) once the
    # user confirms a suggested match, or an exact-name auto-match fires.
    # Null until confirmed — never set from an embedding match alone.
    speaker_id: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True), ForeignKey("speakers.id", ondelete="SET NULL"), nullable=True
    )
    # Set by transcribe_session (routers/processing.py) right after a fresh
    # /embed-speaker call, cleared once /sessions/{id}/speaker-confirm
    # consumes it (see routers/speakers.py). Exists so confirmation doesn't
    # need to re-download the audio and re-run CPU embedding extraction —
    # the embedding is computed once per transcription, used at most once.
    pending_embedding: Mapped[list[float] | None] = mapped_column(
        ARRAY(Float), nullable=True
    )
    location: Mapped[str | None] = mapped_column(String, nullable=True)
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True))
    duration_ms: Mapped[int] = mapped_column(Integer, default=0)

    # Object key in MinIO/S3 (not a full URL — the API mints presigned URLs
    # on demand; see app/services/storage.py), null if audio was discarded.
    audio_object_key: Mapped[str | None] = mapped_column(String, nullable=True)

    locale_id: Mapped[str] = mapped_column(String, default="id_ID")
    status: Mapped[SessionStatus] = mapped_column(
        Enum(SessionStatus, name="session_status"), default=SessionStatus.recorded
    )

    user: Mapped["User"] = relationship(back_populates="sessions")  # noqa: F821
    transcript: Mapped[list["TranscriptSegment"]] = relationship(  # noqa: F821
        back_populates="session",
        cascade="all, delete-orphan",
        order_by="TranscriptSegment.start_ms",
    )
    note: Mapped["KajianNote | None"] = relationship(  # noqa: F821
        back_populates="session", cascade="all, delete-orphan", uselist=False
    )
    speaker_profile: Mapped["Speaker | None"] = relationship(  # noqa: F821
        back_populates="sessions"
    )
